src/js/lib/search.js

```diff
@@ -0,0 +1,344 @@
+    builder = builder.rawOption("post_filter", {
+      bool: {
+        must: [...facetClauses]
+      }
+    })
+
+    // Include suggest if search test is not null/empty
+    if (!emptyOrNil(text)) {
+      builder = builder.rawOption(
+        "suggest",
+        // $FlowFixMe: if we get this far, text is not null
+        buildSuggestQuery(text, LEARN_SUGGEST_FIELDS)
+      )
+    } else if (facetClauses.length === 0) {
+      builder = builder.rawOption("sort", buildDefaultSort())
+    }
+  }
+
+  return builder.build()
+}
+
+export const buildFacetSubQuery = (facets, builder) => {
+  const facetClauses = []
+  if (facets) {
+    facets.forEach((values, key) => {
+      const facetClausesForFacet = []
+
+      if (values && values.length > 0) {
+        addFacetClauseToArray(facetClauses, key, values)
+      }
+
+      // $FlowFixMe: we check for null facets earlier
+      facets.forEach((otherValues, otherKey) => {
+        if (otherKey !== key && otherValues && otherValues.length > 0) {
+          addFacetClauseToArray(facetClausesForFacet, otherKey, otherValues)
+        }
+      })
+
+      if (facetClausesForFacet.length > 0) {
+        const filter = {
+          filter: {
+            bool: {
+              must: [...facetClausesForFacet]
+            }
+          }
+        }
+
+        builder.agg("filter", key, aggregation => {
+          return aggregation
+            .orFilter("bool", filter)
+            .agg(
+              "terms",
+              key === OBJECT_TYPE ? "object_type.keyword" : key,
+              { size: 10000 },
+              key
+            )
+        })
+      } else {
+        builder.agg(
+          "terms",
+          key === OBJECT_TYPE ? "object_type.keyword" : key,
+          { size: 10000 },
+          key
+        )
+      }
+    })
+  }
+  return facetClauses
+}
+
+export const buildOrQuery = (builder, searchType, textQuery, extraClauses) => {
+  const textFilter = emptyOrNil(textQuery) ? [] : [{ bool: textQuery }]
+  builder = builder.orQuery("bool", {
+    filter: {
+      bool: {
+        must: [
+          {
+            term: {
+              object_type: searchType
+            }
+          },
+          ...extraClauses,
+          // Add multimatch text query here to filter out non-matching results
+          ...textFilter
+        ]
+      }
+    },
+    // Add multimatch text query here again to score results based on match
+    ...textQuery
+  })
+  return builder
+}
+
+const addFacetClauseToArray = (facetClauses, facet, values) => {
+  if (
+    facet === OBJECT_TYPE &&
+    values.toString() === buildSearchQuery.toString()
+  ) {
+    return
+  }
+
+  const filterKey = facet === OBJECT_TYPE ? "object_type.keyword" : facet
+
+  facetClauses.push({
+    bool: {
+      should: values.map(value => ({
+        term: {
+          [filterKey]: value
+        }
+      }))
+    }
+  })
+}
+
+export const buildSuggestQuery = (text, suggestFields) => {
```

Comment:
Show the suggestions in a future PR?

Reply:
yeah, I wanted to copy the code over now so I didn't forget